File: Packages/MoonlitCore/Sources/MoonlitCore/IPTV/XtreamClient.swift
// Packages/MoonlitCore/Sources/MoonlitCore/IPTV/XtreamClient.swift
//
// Xtream Codes API client. Swift port of an external IPTV library's Xtream handler.
//
// Xtream panels are notoriously loosely-typed — the same field arrives as a String
// on one server and an Int on another — so responses are walked as untyped JSON
// (`[String: Any]`) via JSONSerialization rather than fought through Codable, which
// matches how the TS treats them as `unknown`.
import Foundation

public enum XtreamError: Error, LocalizedError {
    case auth(String)
    case empty
    case network(String)

    public var errorDescription: String? {
        switch self {
        case .auth(let message): return message
        case .empty: return "The Xtream server returned no channels."
        case .network(let message): return message
        }
    }
}

public enum XtreamContainer: String, Sendable {
    case ts
    case m3u8
}

/// What `fetchUserInfo` learns about a server: which stream containers it allows
/// and the base URL to build stream URLs against (may differ from the API base
/// when the panel reports an HTTPS port).
public struct XtreamServerCaps: Sendable {
    public let allowedFormats: [String]
    public let streamBase: String
}

/// One entry from a channel's short EPG (used by the Phase 2 fallback).
public struct XtreamShortEpgEntry: Sendable {
    public let title: String
    public let description: String?
    public let startMs: Double
    public let endMs: Double
}

public enum XtreamClient {

    /// Base URL + credentials, resolved to the scheme://host[:port] form the API
    /// and stream URLs are built against.
    public struct ResolvedCreds: Sendable {
        public let base: String
        public let username: String
        public let password: String
    }

    private static let userAgent = "IPTVSmartersPro/3.1.5"

    // MARK: - Credential resolution

    /// Resolve user-entered server/username/password into a usable base URL.
    /// Returns nil when the server isn't a valid http(s) URL or creds are blank.
    public static func resolve(_ creds: XtreamCredentials) -> ResolvedCreds? {
        let trimmedServer = creds.server.trimmingCharacters(in: .whitespaces)
            .replacingOccurrences(of: "/+$", with: "", options: .regularExpression)
        let user = creds.username.trimmingCharacters(in: .whitespaces)
        let pass = creds.password.trimmingCharacters(in: .whitespaces)
        guard trimmedServer.range(of: "^https?://", options: [.regularExpression, .caseInsensitive]) != nil,
              !user.isEmpty, !pass.isEmpty,
              let comps = URLComponents(string: trimmedServer),
              let scheme = comps.scheme, let host = comps.host else { return nil }
        let portPart = comps.port.map { ":\($0)" } ?? ""
        return ResolvedCreds(base: "\(scheme)://\(host)\(portPart)", username: user, password: pass)
    }

    /// Extract credentials from a `get.php`/`player_api.php` URL that carries
    /// `username`/`password` query params.
    public static func parseUrl(_ url: String) -> ResolvedCreds? {
        guard let comps = URLComponents(string: url),
              let scheme = comps.scheme, let host = comps.host else { return nil }
        let path = comps.path
        guard path.range(of: "(get\\.php|player_api\\.php)$", options: [.regularExpression, .caseInsensitive]) != nil else { return nil }
        let items = comps.queryItems ?? []
        guard let user = items.first(where: { $0.name == "username" })?.value, !user.isEmpty,
              let pass = items.first(where: { $0.name == "password" })?.value, !pass.isEmpty else { return nil }
        let portPart = comps.port.map { ":\($0)" } ?? ""
        return ResolvedCreds(base: "\(scheme)://\(host)\(portPart)", username: user, password: pass)
    }

    // MARK: - URL building

    private static func apiUrl(_ creds: ResolvedCreds, action: String? = nil, extra: [String: String] = [:]) -> URL? {
        var comps = URLComponents(string: "\(creds.base)/player_api.php")
        var items = [
            URLQueryItem(name: "username", value: creds.username),
            URLQueryItem(name: "password", value: creds.password),
        ]
        if let action { items.append(URLQueryItem(name: "action", value: action)) }
        for (key, value) in extra.sorted(by: { $0.key < $1.key }) {
            items.append(URLQueryItem(name: key, value: value))
        }
        comps?.queryItems = items
        return comps?.url
    }

    /// `{base}/live/{user}/{pass}/{streamId}.{container}`
    public static func buildLiveStreamURL(
        _ creds: ResolvedCreds,
        streamId: String,
        container: XtreamContainer,
        streamBase: String? = nil
    ) -> String {
        let base = streamBase ?? creds.base
        let user = creds.username.addingPercentEncoding(withAllowedCharacters: .urlPathAllowed) ?? creds.username
        let pass = creds.password.addingPercentEncoding(withAllowedCharacters: .urlPathAllowed) ?? creds.password
        return "\(base)/live/\(user)/\(pass)/\(streamId).\(container.rawValue)"
    }

    // MARK: - User info / capabilities

    public static func fetchUserInfo(_ creds: ResolvedCreds) async throws -> XtreamServerCaps {
        guard let url = apiUrl(creds) else { throw XtreamError.auth("Invalid Xtream server URL.") }
        let raw = try await fetchJSON(url)
        guard let dict = raw as? [String: Any],
              let info = dict["user_info"] as? [String: Any] else {
            throw XtreamError.auth("Xtream login did not return account info. Check the server URL.")
        }
        if asInt(info["auth"]) == 0 {
            throw XtreamError.auth("Xtream rejected these credentials (auth failed). Check the server URL, username, and password.")
        }
        switch asString(info["status"])?.lowercased() {
        case "expired": throw XtreamError.auth("This Xtream account is expired.")
        case "banned":  throw XtreamError.auth("This Xtream account is banned by the provider.")
        case "disabled": throw XtreamError.auth("This Xtream account is disabled by the provider.")
        default: break
        }
        let allowedFormats = (info["allowed_output_formats"] as? [Any])?.compactMap { asString($0)?.lowercased() } ?? []
        let streamBase = deriveStreamBase(creds, serverInfo: dict["server_info"] as? [String: Any])
        return XtreamServerCaps(allowedFormats: allowedFormats, streamBase: streamBase)
    }

    private static func deriveStreamBase(_ creds: ResolvedCreds, serverInfo: [String: Any]?) -> String {
        guard let serverInfo,
              asString(serverInfo["server_protocol"])?.lowercased() == "https",
              let comps = URLComponents(string: creds.base), let host = comps.host else {
            return creds.base
        }
        let port = asString(serverInfo["https_port"])?.nonEmpty ?? asString(serverInfo["port"])?.nonEmpty
        return port.map { "https://\(host):\($0)" } ?? "https://\(host)"
    }

    private static func pickContainer(_ pref: XtreamContainer, allowed: [String]) -> XtreamContainer {
        if allowed.isEmpty { return pref }
        if allowed.contains(pref.rawValue) { return pref }
        if allowed.contains("ts") { return .ts }
        if allowed.contains("m3u8") { return .m3u8 }
        return pref
    }

    // MARK: - Live channels

    public static func fetchLiveChannels(
        _ creds: ResolvedCreds,
        baseId: String,
        container: XtreamContainer = .ts,
        caps: XtreamServerCaps? = nil
    ) async throws -> [IPTVChannel] {
        let resolvedContainer = pickContainer(container, allowed: caps?.allowedFormats ?? [])
        let streamBase = caps?.streamBase ?? creds.base
        guard let categoriesURL = apiUrl(creds, action: "get_live_categories"),
              let streamsURL = apiUrl(creds, action: "get_live_streams") else {
            throw XtreamError.network("Could not build Xtream API URLs.")
        }

        async let categoriesRaw = fetchJSON(categoriesURL)
        async let streamsRaw = fetchJSON(streamsURL)
        let (categories, streams) = try await (categoriesRaw, streamsRaw)

        var categoryName: [String: String] = [:]
        if let rows = categories as? [[String: Any]] {
            for row in rows {
                if let id = asString(row["category_id"]) {
                    categoryName[id] = asString(row["category_name"]) ?? ""
                }
            }
        }

        guard let streamRows = streams as? [[String: Any]] else { return [] }
        var out: [IPTVChannel] = []
        out.reserveCapacity(streamRows.count)
        for row in streamRows {
            guard let streamId = asString(row["stream_id"]) else { continue }
            let tvgId = asString(row["epg_channel_id"])?.trimmingCharacters(in: .whitespaces).nonEmpty
            let group = asString(row["category_id"]).flatMap { categoryName[$0] }?.nonEmpty
            let url = buildLiveStreamURL(creds, streamId: streamId, container: resolvedContainer, streamBase: streamBase)

            var catchupType: String?
            var catchupDays: Int?
            if let archive = asInt(row["tv_archive"]), archive > 0 {
                catchupType = "xtream"
                if let days = asInt(row["tv_archive_duration"]), days > 0 { catchupDays = days }
            }

            let name = asString(row["name"])?.trimmingCharacters(in: .whitespaces).nonEmpty ?? "Stream \(streamId)"
            let logo = asString(row["stream_icon"])?.trimmingCharacters(in: .whitespaces).nonEmpty

            out.append(
                IPTVChannel(
                    id: "\(baseId)::xt::\(streamId)",
                    tvgId: tvgId,
                    name: name,
                    logo: logo,
                    group: group,
                    url: url,
                    catchupType: catchupType,
                    catchupDays: catchupDays,
                    headers: nil
                )
            )
        }
        return out
    }

    // MARK: - Short EPG (Phase 2 fallback)

    public static func fetchShortEpg(_ creds: ResolvedCreds, streamId: String, limit: Int = 8) async throws -> [XtreamShortEpgEntry] {
        guard let url = apiUrl(creds, action: "get_short_epg", extra: ["stream_id": streamId, "limit": String(limit)]) else {
            return []
        }
        let raw = (try? await fetchJSON(url)) ?? [:]
        guard let dict = raw as? [String: Any],
              let listings = dict["epg_listings"] as? [[String: Any]] else { return [] }
        var out: [XtreamShortEpgEntry] = []
        for row in listings {
            guard let start = asDouble(row["start_timestamp"]), let stop = asDouble(row["stop_timestamp"]) else { continue }
            let startMs = start * 1000, endMs = stop * 1000
            guard endMs > startMs else { continue }
            out.append(
                XtreamShortEpgEntry(
                    title: decodeBase64(asString(row["title"])).nonEmpty ?? "Untitled",
                    description: decodeBase64(asString(row["description"])).nonEmpty,
                    startMs: startMs,
                    endMs: endMs
                )
            )
        }
        return out
    }

    // MARK: - Networking

    private static func fetchJSON(_ url: URL) async throws -> Any {
        var request = URLRequest(url: url)
        request.setValue(userAgent, forHTTPHeaderField: "User-Agent")
        request.timeoutInterval = 20
        do {
            let (data, response) = try await URLSession.shared.data(for: request)
            if let http = response as? HTTPURLResponse, !(200..<300).contains(http.statusCode) {
                throw XtreamError.network("Xtream server returned HTTP \(http.statusCode).")
            }
            return try JSONSerialization.jsonObject(with: data, options: [])
        } catch let error as XtreamError {
            throw error
        } catch {
            throw XtreamError.network(error.localizedDescription)
        }
    }

    // MARK: - Loose JSON accessors

    private static func asString(_ value: Any?) -> String? {
        switch value {
        case let s as String: return s
        case let i as Int: return String(i)
        case let d as Double: return d == d.rounded() ? String(Int(d)) : String(d)
        case let b as Bool: return b ? "1" : "0"
        default: return nil
        }
    }

    private static func asInt(_ value: Any?) -> Int? {
        switch value {
        case let i as Int: return i
        case let d as Double: return Int(d)
        case let s as String: return Int(s) ?? Double(s).map(Int.init)
        case let b as Bool: return b ? 1 : 0
        default: return nil
        }
    }

    private static func asDouble(_ value: Any?) -> Double? {
        switch value {
        case let d as Double: return d
        case let i as Int: return Double(i)
        case let s as String: return Double(s)
        default: return nil
        }
    }

    /// Base64-decode Xtream's title/description fields, but only when the input
    /// really looks like base64 and decodes to clean UTF-8 — otherwise return it
    /// unchanged (many panels send plain text here).
    private static func decodeBase64(_ input: String?) -> String {
        guard let raw = input?.trimmingCharacters(in: .whitespaces), !raw.isEmpty else { return "" }
        guard raw.range(of: "^[A-Za-z0-9+/]+={0,2}$", options: .regularExpression) != nil,
              raw.count % 4 == 0,
              let data = Data(base64Encoded: raw),
              let decoded = String(data: data, encoding: .utf8) else { return raw }
        let trimmed = decoded.trimmingCharacters(in: .whitespaces)
        if trimmed.isEmpty || hasControlChars(trimmed) { return raw }
        return trimmed
    }

    private static func hasControlChars(_ s: String) -> Bool {
        for scalar in s.unicodeScalars {
            let v = scalar.value
            if v < 0x20, v != 0x09, v != 0x0A, v != 0x0D { return true }
        }
        return false
    }
}

private extension String {
    var nonEmpty: String? { isEmpty ? nil : self }
}

File: moonlit-web/src/lib/iptv/xtream-client.ts
import type { IPTVChannel, XtreamCredentials, XtreamServerCaps, XtreamShortEpgEntry } from './models';

const USER_AGENT = 'IPTVSmartersPro/3.1.5';

export class XtreamError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'XtreamError';
  }
}

export interface ResolvedCreds {
  base: string;
  username: string;
  password: string;
}

function asString(value: unknown): string | undefined {
  if (typeof value === 'string') return value;
  if (typeof value === 'number') return Number.isInteger(value) ? String(value) : String(value);
  if (typeof value === 'boolean') return value ? '1' : '0';
  return undefined;
}

function asInt(value: unknown): number | undefined {
  if (typeof value === 'number') return Math.floor(value);
  if (typeof value === 'string') {
    const n = parseInt(value, 10);
    if (!isNaN(n)) return n;
    const f = parseFloat(value);
    if (!isNaN(f)) return Math.floor(f);
  }
  if (typeof value === 'boolean') return value ? 1 : 0;
  return undefined;
}

function asDouble(value: unknown): number | undefined {
  if (typeof value === 'number') return value;
  if (typeof value === 'string') {
    const f = parseFloat(value);
    if (!isNaN(f)) return f;
  }
  return undefined;
}

function nonEmpty(s: string | undefined | null): string | undefined {
  if (!s) return undefined;
  return s.trim() || undefined;
}

function decodeBase64(input: string | undefined): string {
  if (!input) return '';
  const raw = input.trim();
  if (!raw) return '';
  if (!/^[A-Za-z0-9+/]+={0,2}$/.test(raw) || raw.length % 4 !== 0) return raw;
  try {
    const decoded = atob(raw);
    const trimmed = decoded.trim();
    if (!trimmed) return raw;
    for (let i = 0; i < trimmed.length; i++) {
      const c = trimmed.charCodeAt(i);
      if (c < 0x20 && c !== 0x09 && c !== 0x0a && c !== 0x0d) return raw;
    }
    return trimmed;
  } catch {
    return raw;
  }
}

async function fetchJSON(url: string): Promise<unknown> {
  const controller = new AbortController();
  const timeout = setTimeout(() => controller.abort(), 20_000);
  try {
    const response = await fetch(url, {
      headers: { 'User-Agent': USER_AGENT },
      signal: controller.signal,
    });
    if (!response.ok) throw new XtreamError(`Xtream server returned HTTP ${response.status}.`);
    return await response.json();
  } catch (err) {
    if (err instanceof XtreamError) throw err;
    throw new XtreamError(err instanceof Error ? err.message : 'Network error');
  } finally {
    clearTimeout(timeout);
  }
}

function apiUrl(
  creds: ResolvedCreds,
  action?: string,
  extra: Record<string, string> = {},
): string {
  const url = new URL(`${creds.base}/player_api.php`);
  url.searchParams.set('username', creds.username);
  url.searchParams.set('password', creds.password);
  if (action) url.searchParams.set('action', action);
  for (const [key, value] of Object.entries(extra).sort(([a], [b]) => a.localeCompare(b))) {
    url.searchParams.set(key, value);
  }
  return url.toString();
}

function deriveStreamBase(creds: ResolvedCreds, serverInfo: unknown): string {
  const info = serverInfo as Record<string, unknown> | undefined;
  if (!info) return creds.base;
  const protocol = asString(info['server_protocol'])?.toLowerCase();
  if (protocol !== 'https') return creds.base;
  const host = new URL(creds.base).hostname;
  const port = nonEmpty(asString(info['https_port'])) || nonEmpty(asString(info['port']));
  return port ? `https://${host}:${port}` : `https://${host}`;
}

function pickContainer(pref: 'ts' | 'm3u8', allowed: string[]): 'ts' | 'm3u8' {
  if (!allowed.length) return pref;
  if (allowed.includes(pref)) return pref;
  if (allowed.includes('ts')) return 'ts';
  if (allowed.includes('m3u8')) return 'm3u8';
  return pref;
}

export function resolveCreds(creds: XtreamCredentials): ResolvedCreds | null {
  const trimmedServer = creds.server.trim().replace(/\/+$/, '');
  const user = creds.username.trim();
  const pass = creds.password.trim();
  if (!/^https?:\/\//i.test(trimmedServer) || !user || !pass) return null;
  let url: URL;
  try { url = new URL(trimmedServer); } catch { return null; }
  return { base: `${url.protocol}//${url.host}${url.port ? `:${url.port}` : ''}`, username: user, password: pass };
}

export function parseUrl(urlString: string): ResolvedCreds | null {
  let comps: URL;
  try { comps = new URL(urlString); } catch { return null; }
  const path = comps.pathname;
  if (!/(\bget\.php\b|\bplayer_api\.php\b)$/i.test(path)) return null;
  const user = comps.searchParams.get('username');
  const pass = comps.searchParams.get('password');
  if (!user || !pass) return null;
  return { base: `${comps.protocol}//${comps.host}${comps.port ? `:${comps.port}` : ''}`, username: user, password: pass };
}

export function buildLiveStreamURL(
  creds: ResolvedCreds,
  streamId: string,
  container: 'ts' | 'm3u8',
  streamBase?: string,
): string {
  const base = streamBase || creds.base;
  const user = encodeURIComponent(creds.username);
  const pass = encodeURIComponent(creds.password);
  return `${base}/live/${user}/${pass}/${streamId}.${container}`;
}

export async function fetchUserInfo(creds: ResolvedCreds): Promise<XtreamServerCaps> {
  const url = apiUrl(creds);
  const raw = await fetchJSON(url);
  const dict = raw as Record<string, unknown>;
  const info = dict['user_info'] as Record<string, unknown> | undefined;
  if (!info) throw new XtreamError('Xtream login did not return account info. Check the server URL.');
  if (asInt(info['auth']) === 0) throw new XtreamError('Xtream rejected these credentials (auth failed).');
  const status = asString(info['status'])?.toLowerCase();
  if (status === 'expired') throw new XtreamError('This Xtream account is expired.');
  if (status === 'banned') throw new XtreamError('This Xtream account is banned.');
  if (status === 'disabled') throw new XtreamError('This Xtream account is disabled.');
  const allowedFormats = ((info['allowed_output_formats'] as unknown[]) ?? [])
    .map(f => asString(f)?.toLowerCase())
    .filter(Boolean) as string[];
  const streamBase = deriveStreamBase(creds, dict['server_info']);
  return { allowedFormats, streamBase };
}

export async function fetchLiveChannels(
  creds: ResolvedCreds,
  baseId: string,
  container: 'ts' | 'm3u8' = 'ts',
  caps?: XtreamServerCaps,
): Promise<IPTVChannel[]> {
  const resolvedContainer = pickContainer(container, caps?.allowedFormats ?? []);
  const streamBase = caps?.streamBase ?? creds.base;
  const categoriesURL = apiUrl(creds, 'get_live_categories');
  const streamsURL = apiUrl(creds, 'get_live_streams');

  const [categoriesRaw, streamsRaw] = await Promise.all([
    fetchJSON(categoriesURL),
    fetchJSON(streamsURL),
  ]);

  const categoryName: Record<string, string> = {};
  if (Array.isArray(categoriesRaw)) {
    for (const row of categoriesRaw as Record<string, unknown>[]) {
      const id = asString(row['category_id']);
      if (id) categoryName[id] = asString(row['category_name']) ?? '';
    }
  }

  if (!Array.isArray(streamsRaw)) return [];
  const out: IPTVChannel[] = [];
  for (const row of streamsRaw as Record<string, unknown>[]) {
    const streamId = asString(row['stream_id']);
    if (!streamId) continue;
    const tvgId = nonEmpty(asString(row['epg_channel_id'])) || undefined;
    const catId = asString(row['category_id']);
    const group = catId && categoryName[catId] ? categoryName[catId] : undefined;
    const url = buildLiveStreamURL(creds, streamId, resolvedContainer, streamBase);

    let catchupType: string | undefined;
    let catchupDays: number | undefined;
    const archive = asInt(row['tv_archive'] as unknown);
    if (archive && archive > 0) {
      catchupType = 'xtream';
      const days = asInt(row['tv_archive_duration'] as unknown);
      if (days && days > 0) catchupDays = days;
    }

    const name = nonEmpty(asString(row['name'])) || `Stream ${streamId}`;
    const logo = nonEmpty(asString(row['stream_icon'])) || undefined;

    out.push({
      id: `${baseId}::xt::${streamId}`,
      tvgId,
      name,
      logo,
      group,
      url,
      catchupType,
      catchupDays,
    });
  }
  return out;
}

export async function fetchShortEpg(
  creds: ResolvedCreds,
  streamId: string,
  limit: number = 8,
): Promise<XtreamShortEpgEntry[]> {
  const url = apiUrl(creds, 'get_short_epg', { stream_id: streamId, limit: String(limit) });
  let raw: unknown;
  try { raw = await fetchJSON(url); } catch { return []; }
  const dict = raw as Record<string, unknown> | undefined;
  if (!dict) return [];
  const listings = dict['epg_listings'] as Record<string, unknown>[] | undefined;
  if (!listings) return [];
  const out: XtreamShortEpgEntry[] = [];
  for (const row of listings) {
    const start = asDouble(row['start_timestamp']);
    const stop = asDouble(row['stop_timestamp']);
    if (!start || !stop) continue;
    const startMs = start * 1000;
    const endMs = stop * 1000;
    if (endMs <= startMs) continue;
    out.push({
      title: decodeBase64(asString(row['title'])) || 'Untitled',
      description: nonEmpty(decodeBase64(asString(row['description']))) || undefined,
      startMs,
      endMs,
    });
  }
  return out;
}
